library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,44 @@
+[JLBP-4] Avoid dependencies on unstable libraries and features
+--------------------------------------------------------------
+
+- Unstable libraries are defined as libraries that allow for breaking changes to
+  their Public API within the same major version. For libraries following
+  semver, this means libraries with a 0.x.y version. (See [JLBP-3](JLBP-3.md)
+  for more details on the recommendations for following semver.)
+- Unstable features are defined as features that are not part of the stable
+  Public API of a stable library, often marked with annotations like `@Beta`.
+  (See [JLBP-3](JLBP-3.md) for more details on the recommendations for
+  annotating unstable features.)
+- If your library depends on an unstable library or feature, and that feature
+  experiences a breaking change between versions, your library will be locked to
+  a specific version of your dependency.
+  - If you expose the unstable feature on your library's surface, then your
+    library's current major version will be permanently locked to the version
+    you initially exposed, and you won't be able to upgrade without making a
+    breaking change to your users.
+  - If you only use the unstable feature in your implementation, then each minor
+    or patch version of your library will require a very specific version of
+    your dependency, and it will be unsafe for your users to upgrade your
+    library on its own, creating opportunities for hard-to-diagnose runtime
+    conflicts for users.
+- Given the consequences of depending unstable features in dependencies, avoid
+  doing so.
+  - Depending on unstable features between submodules of a single library is
+    considered acceptable, given there is a way for users to ensure they are
```

Comment:
This might be nitpicking of phrasing, but: Using the same version number for all modules (or, likewise, providing a BOM) doesn't "ensure" much. It just means that, once users recognize that there's a problem, they have a relatively easy way to fix it.

I do think we'll want to do some work to help users detect these problems in the first place -- possibly by following [gRPC's lead](https://repo1.maven.org/maven2/io/grpc/grpc-auth/1.14.0/grpc-auth-1.14.0.pom) in using single-element version ranges like `[1.14.0]`. (We've at least casually referred to this option internally, but I have yet to flesh it out. Hopefully later today...)

Reply:
Agreed, we don't need to wait to finalize this item (or at least to create it -- we can always edit later :))

For our current purposes, how would you feel about phrasing like...

  - Depending on unstable features between submodules of a single library is acceptable, provided that users can force their build system to use compatible versions of the submodules easily, e.g. by having the library use the same version number for all submodules, or by providing a BOM.

Reply:
I will use a variant of what you proposed.